pkg/kubelet/kubelet_pods.go

```diff
@@ -1224,33 +1246,27 @@ func (pk *podKillerWithChannel) KillPod(pair *kubecontainer.PodPair) {
 // PerformPodKillingWork launches a goroutine to kill a pod received from the channel if
 // another goroutine isn't already in action.
 func (pk *podKillerWithChannel) PerformPodKillingWork() {
-	killing := sets.NewString()
-	// guard for the killing set
-	lock := sync.Mutex{}
 	for podPair := range pk.podKillingCh {
 		runningPod := podPair.RunningPod
 		apiPod := podPair.APIPod
 
-		lock.Lock()
-		exists := killing.Has(string(runningPod.ID))
-		if !exists {
-			killing.Insert(string(runningPod.ID))
+		if pk.IsPodPendingTerminationByUID(runningPod.ID) {
+			continue
 		}
-		lock.Unlock()
 
-		if !exists {
-			go func(apiPod *v1.Pod, runningPod *kubecontainer.Pod) {
-				klog.V(2).Infof("Killing unwanted pod %q", runningPod.Name)
-				err := pk.killPod(apiPod, runningPod, nil, nil)
-				if err != nil {
-					klog.Errorf("Failed killing the pod %q: %v", runningPod.Name, err)
-				}
-				lock.Lock()
-				killing.Delete(string(runningPod.ID))
-				lock.Unlock()
-				pk.markMirrorPodTerminated(string(runningPod.ID))
-			}(apiPod, runningPod)
-		}
+		pk.podKillingLock.Lock()
+		pk.podTerminationMap[string(runningPod.ID)] = runningPod.Name
+		pk.podKillingLock.Unlock()
```

Comment:
I don't really know these code paths, but why is this not an atomic operation? Adding runningPod.Name could be part of the call to IsPodPendingTerminationByUID (it would probably need to be named something different, `CheckAndMarkPodPendingTerminationByUID`?), so no other entity would have a chance to race and sneak in between these two locks

Reply:
good idea... updated